### PR TITLE
Ensure Redis Proxy can follow redirection when doing request mirroring

### PR DESCRIPTION
**Pull request by HenryYYang**
**Description**: Currently the Redis redirection logic is part of the command splitter, this has the consequence that shadow traffic to other clusters are not able to follow the redirection responses.  Part of the challenge of the change is that the original request is the storage of the original request, as it could be destroyed before the redirection command is received.

The solution is refactoring the redirection logic into the Redis ConnPool.  This greatly simplify the redirection logic.  Also the connection pool is the only entity that can meaningfully manage the connections to upstream and correctly drain the connections.  This also require storing a shared_ptr to the incoming request with the connection pool.
Risk Level: High
Testing: Add unit tests
Docs Changes: Updated
Release Notes: Updated

### CODE REVIEW

**Comment by jmarantz** [comment]
I took the liberty of adding some markdown quotes to your PR description so the table is legible.

But below the table can you summarize your conclusions from it?

**Comment by jmarantz** [comment]
master merge needed.

**Comment by jmarantz** [comment]
drive by comment from description: does it make sense to flag-control this performance trade-off, as you indicated there are cases where the new approach is slower. Maybe that describes someone's use-case for which this is a regression?

**Comment by HenryYYang** [comment]
I think most of the cost is from creating the new shared_ptr, I think I can work around this and improve the performance.  I'll report back.

**Comment by HenryYYang** [comment]
Ok, I've update the code and the microbenchmark result.  It's looking much better.

**Comment by mattklein123** [comment]
Putting this into wait until we get the redis test flake fixed and master merged into this.

/wait

**Comment by HenryYYang** [comment]
@msukalski @FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg PTAL

File: source/extensions/filters/network/common/redis/utility.cc [hunk 1]
```diff
@@ -19,6 +19,36 @@ AuthRequest::AuthRequest(const std::string& password) {
...
+    return true;
+  }
+  error_substrings = StringUtil::splitToken(error_response.asString(), " ", false);
+  if (error_substrings.size() != 3) {
+    return true;
+  }
+  if (error_substrings[0] == "ASK") {
+    ask_redirection = true;
+  } else if (error_substrings[0] == "MOVED") {
+    ask_redirection = false;
```
[the remark this answers is not in the document]
Comment (HenryYYang) [reply source/extensions/filters/network/common/redis/utility.cc]: This is from a refactored code, but as it turned out this is only called from a single place now(there used to be multiple callers).  And more importantly, in that code path, we already validate for these parameters in the client, we should just pass them in.  I'm going to remove this function.

File: source/extensions/filters/network/common/redis/client_impl.cc [hunk 2]
```diff
@@ -224,17 +224,26 @@ void ClientImpl::onRespValue(RespValuePtr&& value) {
...
-        if (redirected) {
-          host_->cluster().stats().upstream_internal_redirect_succeeded_total_.inc();
-        } else {
-          host_->cluster().stats().upstream_internal_redirect_failed_total_.inc();
-        }
+      // MOVED and ASK redirection errors have the following substrings: MOVED or ASK (err[0]), hash
+      // key slot (err[1]), and IP address and TCP port separated by a colon (err[2])
+      if (err[0] == RedirectionResponse::get().MOVED) {
+        redirected = true;
+        redirect_succeeded = callbacks.onRedirection(std::move(value), std::string(err[2]), false);
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline source/extensions/filters/network/common/redis/client_impl.cc]: nit: might be easier to do a ternary operator on MOVED/ASK for the last param, instead of repeating :232-233 again as :235-236

File: test/extensions/filters/network/common/redis/mocks.cc [hunk 3]
```diff
@@ -42,16 +42,24 @@ MockClient::MockClient() {
       }));
   ON_CALL(*this, close()).WillByDefault(Invoke([this]() -> void {
     raiseEvent(Network::ConnectionEvent::LocalClose);
+    //    for (ClientCallbacks * callbacks : client_callbacks_) {
+    //      callbacks->onFailure();
+    //    }
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline test/extensions/filters/network/common/redis/mocks.cc]: Do you want to keep the commented out stuff?

File: source/extensions/filters/network/redis_proxy/command_splitter_impl.cc [hunk 4]
```diff
@@ -153,18 +110,17 @@ SplitRequestPtr SimpleRequest::create(Router& router,
 
   const auto route = router.upstreamPool(incoming_request->asArray()[1].asString());
   if (route) {
-    request_ptr->conn_pool_ = route->upstream();
+    Common::Redis::RespValueSharedPtr base_request = std::move(incoming_request);
     request_ptr->handle_ =
-        makeRequest(route, incoming_request->asArray()[0].asString(),
-                    incoming_request->asArray()[1].asString(), *incoming_request, *request_ptr);
+        makeSingleServerRequest(route, base_request->asArray()[0].asString(),
+                                base_request->asArray()[1].asString(), base_request, *request_ptr);
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline source/extensions/filters/network/redis_proxy/command_splitter_impl.cc]: possible simplication idea- do `base_request->asArray()[X].asString()` inside `makeSingleServerRequest`?

Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply source/extensions/filters/network/redis_proxy/command_splitter_impl.cc]: I guess that doesn't work though with the way that we have commands and keys for `EVAL` though...

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc [hunk 5]
```diff
@@ -97,16 +97,54 @@ class RedisConnPoolImplTest : public testing::Test, public Common::Redis::Client
...
+            }));
     Common::Redis::Client::PoolRequest* request =
         conn_pool_->makeRequest(hash_key, value, callbacks);
-    EXPECT_EQ(&active_request, request);
+    EXPECT_NE(nullptr, request);
+    EXPECT_NE(nullptr, client_callbacks.back());
+
+    EXPECT_CALL(active_request, cancel());
+    request->cancel();
+    //    Common::Redis::RespValuePtr resp_value = std::make_unique<Common::Redis::RespValue>();
```
Comment (mattklein123) [inline test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc]: del commented out code, here and elsewhere.

File: test/extensions/filters/network/redis_proxy/command_split_speed_test.cc [hunk 6]
```diff
@@ -43,6 +43,23 @@ class CommandSplitSpeedTest {
 
     return request;
   }
+  using ValueOrPointer =
+      absl::variant<const Common::Redis::RespValue, Common::Redis::RespValueConstSharedPtr>;
+
+  void createShared(Common::Redis::RespValueSharedPtr request) {
+    for (uint64_t i = 1; i < request->asArray().size(); i += 2) {
+      auto single_set = std::make_shared<const Common::Redis::RespValue>(
+          request, Common::Redis::Utility::SetRequest::instance(), i, i + 2);
```
Comment (antoniovicente) [inline test/extensions/filters/network/redis_proxy/command_split_speed_test.cc]: "i + 2" looks like a typo.  See #9541
